File: microservices/UI/src/Dtos/experimentReportDto.ts
import {ArtefactDto} from "./artefactDto";

export class StateRequestDto {
  state: string | undefined;
  version: string | undefined;
  secure: boolean | undefined;
  dirty: boolean | undefined;
  experimentReportDto : string | undefined | ExperimentReportDto
}

export class ExperimentReportDto{
  Timestamp: string | undefined;
  RunningInstances: ArtefactDto[]| undefined;
  AverageLatency: number| undefined;
  Stability: number | undefined
  ErrorCount: number| undefined;
  Message: string| undefined;
}

File: microservices/NodeController/ApplicationLogic/OutGoing/ExperimentReportDto.cs
using ApplicationLogic.Interfaces.Dao;
using Entities.DomainInterfaces.Planned;
using Entities.DomainInterfaces.ResourceManagement;

namespace ApplicationLogic.OutGoing;

public class ExperimentReportDto
{
    public DateTime Timestamp { get; set; }

    public List<ArtifactDto> RunningInstances { get; set; }

    public double AverageLatency { get; set; }

    public int ErrorCount { get; set; }

    public string Message { get; set; }

    public static ExperimentReportDto TryConvertToDto(IExperimentReport experimentReport, IPodReadService podReadService)
    {
        return new ExperimentReportDto
        {
            Timestamp = experimentReport.Timestamp,
            AverageLatency = Math.Round(experimentReport.ExperimentResult?.AverageLatency ?? -1.0, 3),
            ErrorCount = experimentReport.ExperimentResult?.ErrorCount ?? default,
            Message = experimentReport.ExperimentResult?.Message ?? string.Empty,
            RunningInstances = experimentReport.RunningInstances
                .ConvertAll(serviceInstance => ArtifactDto.TryConvertToDto(serviceInstance, GetServiceType(serviceInstance.Type, podReadService)))
                                                                    .Where(artifactDto => artifactDto is not null).ToList()!
        };
    }

    private static IServiceType? GetServiceType(string serviceType, IPodReadService podReadService) => podReadService.GetServiceType(serviceType);
}